spirv_glsl.cpp

```diff
@@ -719,94 +719,51 @@ void CompilerGLSL::emit_header()
 			statement("#extension ", ext, " : enable");
 			statement("#endif");
 		}
-		else if (ext == "GL_KHR_shader_subgroup_ballot")
-		{
-			statement("#ifdef GL_KHR_shader_subgroup");
-			statement("#extension GL_KHR_shader_subgroup : require");
-
-			statement("#elif defined(GL_NV_shader_thread_group) && defined(GL_NV_shader_thread_shuffle)");
-			statement("#extension GL_NV_shader_thread_group : require");
-			statement("#extension GL_NV_shader_thread_shuffle : require");
-			statement("#define gl_SubgroupEqMask uvec4(gl_ThreadEqMaskNV, 0u, 0u, 0u)");
-			statement("#define gl_SubgroupGeMask uvec4(gl_ThreadGeMaskNV, 0u, 0u, 0u)");
-			statement("#define gl_SubgroupGtMask uvec4(gl_ThreadGtMaskNV, 0u, 0u, 0u)");
-			statement("#define gl_SubgroupLeMask uvec4(gl_ThreadLeMaskNV, 0u, 0u, 0u)");
-			statement("#define gl_SubgroupLtMask uvec4(gl_ThreadLtMaskNV, 0u, 0u, 0u)");
-
-			statement("#elif defined(GL_ARB_shader_ballot) && defined(GL_ARB_gpu_shader_int64)");
-			statement("#extension GL_ARB_shader_ballot : require");
-			statement("#extension GL_ARB_gpu_shader_int64 : require");
-			statement("#define gl_SubgroupEqMask uvec4(unpackUint2x32(gl_SubGroupEqMaskARB), 0u, 0u)");
-			statement("#define gl_SubgroupGeMask uvec4(unpackUint2x32(gl_SubGroupGeMaskARB), 0u, 0u)");
-			statement("#define gl_SubgroupGtMask uvec4(unpackUint2x32(gl_SubGroupGtMaskARB), 0u, 0u)");
-			statement("#define gl_SubgroupLeMask uvec4(unpackUint2x32(gl_SubGroupLeMaskARB), 0u, 0u)");
-			statement("#define gl_SubgroupLtMask uvec4(unpackUint2x32(gl_SubGroupLtMaskARB), 0u, 0u)");
+		else
+			statement("#extension ", ext, " : require");
+	}
 
-			statement("#else");
-			statement("#error No extensions available to emulate GL_KHR_shader_subgroup_ballot.");
+	if (!options.vulkan_semantics)
+	{
+		using Supp = KHR_shader_subgroupSupportHelper;
 
-			statement("#endif");
-		}
-		else if (ext == "GL_KHR_shader_subgroup_vote")
+		Supp::Result result = shader_subgroup_supporter.resolve();
+		for (uint32_t ft = 0u; ft < Supp::FeatureCount; ++ft)
 		{
-			declared_KHR_shader_subgroup_vote = true;
-			if (options.vulkan_semantics)
-				statement("#extension GL_KHR_shader_subgroup_vote : require");
-			else
+			Supp::Feature feature = static_cast<Supp::Feature>(ft);
+			if (!shader_subgroup_supporter.isFeatureRequested(feature))
+				continue;
+
+			auto exts = shader_subgroup_supporter.getCandidatesForFeature(feature, result);
+			if (exts.empty())
+				continue;
```

Comment:
wouldn't that require you to emit an `#error` statement?

Reply:
no, empty candidates list means no extension is needed to emulate

Reply:
ah ok, better document it.